File: packages/domain/validation.py
"""Deterministic validation rules. Pure functions, no LLM, no I/O."""

from __future__ import annotations

from collections.abc import Collection, Mapping, Sequence
from datetime import date
from decimal import ROUND_HALF_UP, Decimal, InvalidOperation
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

REQUIRED_DOCS_RULE = "REQUIRED_DOCS"
POLICY_ACTIVE_RULE = "POLICY_ACTIVE"
FIELD_CONSISTENCY_RULE = "FIELD_CONSISTENCY"
DATE_LOGIC_RULE = "DATE_LOGIC"
AMOUNT_RECONCILIATION_RULE = "AMOUNT_RECONCILIATION"
DUPLICATE_SHA256_RULE = "DUPLICATE_SHA256"

DEFAULT_REQUIRED_DOCS: tuple[str, ...] = ("CLAIM_FORM", "HOSPITAL_FINAL_BILL", "DISCHARGE_SUMMARY")
_CENT = Decimal("0.01")


class Severity(StrEnum):
    PASS = "PASS"
    WARNING = "WARNING"
    FAIL = "FAIL"
    BLOCK = "BLOCK"


class ValidationResult(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")
    rule_id: str
    passed: bool
    severity: Severity
    evidence_ids: list[str] = Field(default_factory=list)
    message: str = ""


def _ok(rule_id: str, evidence_ids: Sequence[str], message: str = "") -> ValidationResult:
    return ValidationResult(
        rule_id=rule_id,
        passed=True,
        severity=Severity.PASS,
        evidence_ids=list(evidence_ids),
        message=message,
    )


def _bad(
    rule_id: str, evidence_ids: Sequence[str], message: str, severity: Severity = Severity.FAIL
) -> ValidationResult:
    return ValidationResult(
        rule_id=rule_id,
        passed=False,
        severity=severity,
        evidence_ids=list(evidence_ids),
        message=message,
    )


def validate_required_docs(
    present_types: Collection[str],
    required_types: Collection[str] = DEFAULT_REQUIRED_DOCS,
    evidence_ids: Sequence[str] = (),
) -> ValidationResult:
    present = {str(t).strip().upper() for t in present_types if str(t).strip()}
    required = [str(t).strip().upper() for t in required_types if str(t).strip()]
    missing = [t for t in required if t not in present]
    if not missing:
        return _ok(REQUIRED_DOCS_RULE, evidence_ids, "All required documents present.")
    return _bad(REQUIRED_DOCS_RULE, evidence_ids, f"Missing: {', '.join(missing)}.")


def validate_policy_active(
    incident_date: date | None,
    effective_from: date | None,
    effective_to: date | None,
    evidence_ids: Sequence[str] = (),
) -> ValidationResult:
    if incident_date is None or effective_from is None:
        return _bad(
            POLICY_ACTIVE_RULE, evidence_ids, "Incident/policy start missing.", Severity.BLOCK
        )
    if incident_date < effective_from:
        return _bad(
            POLICY_ACTIVE_RULE, evidence_ids, "Incident precedes policy start.", Severity.BLOCK
        )
    if effective_to is not None and incident_date > effective_to:
        return _bad(POLICY_ACTIVE_RULE, evidence_ids, "Incident after policy end.", Severity.BLOCK)
    return _ok(POLICY_ACTIVE_RULE, evidence_ids, "Policy active on incident date.")


def _norm(v: str | None) -> str:
    return " ".join(str(v).strip().casefold().split())


def validate_field_consistency(
    field_name: str, values_by_source: Mapping[str, str | None], evidence_ids: Sequence[str] = ()
) -> ValidationResult:
    distinct = {_norm(v) for v in values_by_source.values() if v and _norm(v)}
    if not distinct:
        return _bad(FIELD_CONSISTENCY_RULE, evidence_ids, f"'{field_name}': no values.")
    if len(distinct) == 1:
        return _ok(FIELD_CONSISTENCY_RULE, evidence_ids, f"'{field_name}' consistent.")
    return _bad(FIELD_CONSISTENCY_RULE, evidence_ids, f"'{field_name}' mismatch.")


def validate_date_logic(
    admission_date: date | None, discharge_date: date | None, evidence_ids: Sequence[str] = ()
) -> ValidationResult:
    if admission_date is None or discharge_date is None:
        return _bad(DATE_LOGIC_RULE, evidence_ids, "Admission/discharge missing.")
    if admission_date > discharge_date:
        return _bad(DATE_LOGIC_RULE, evidence_ids, "Admission after discharge.")
    return _ok(DATE_LOGIC_RULE, evidence_ids, "Date logic holds.")


def _money(v: Decimal | int | str) -> Decimal | None:
    try:
        return Decimal(str(v)).quantize(_CENT, rounding=ROUND_HALF_UP)
    except (InvalidOperation, ValueError, TypeError):
        return None


def validate_amount_reconciliation(
    gross: Decimal | int | str,
    discount: Decimal | int | str,
    net: Decimal | int | str,
    evidence_ids: Sequence[str] = (),
    tolerance: Decimal = _CENT,
) -> ValidationResult:
    g, d, n = _money(gross), _money(discount), _money(net)
    if g is None or d is None or n is None:
        return _bad(AMOUNT_RECONCILIATION_RULE, evidence_ids, "Unparseable amount.")
    if g < 0 or d < 0 or n < 0:
        return _bad(AMOUNT_RECONCILIATION_RULE, evidence_ids, "Negative amount.")
    if abs((g - d) - n) <= abs(tolerance):
        return _ok(AMOUNT_RECONCILIATION_RULE, evidence_ids, f"Reconciled: {g}-{d}=={n}.")
    return _bad(AMOUNT_RECONCILIATION_RULE, evidence_ids, f"Expected {(g - d):.2f}, got {n:.2f}.")


def validate_duplicate_sha256(
    sha256_list: Sequence[str], evidence_ids: Sequence[str] = ()
) -> ValidationResult:
    seen: set[str] = set()
    for h in sha256_list:
        key = str(h).strip().lower()
        if not key:
            continue
        if key in seen:
            return _bad(DUPLICATE_SHA256_RULE, evidence_ids, "Duplicate sha256.")
        seen.add(key)
    return _ok(DUPLICATE_SHA256_RULE, evidence_ids, "No duplicates.")

File: apps/api/internal/validation/validation.go
// Package validation implements the deterministic claim validation core.
//
// It is the Go port of packages/domain/validation.py and satisfies the
// behavioral contract in fixtures/golden_cases.json (cases 01-06).
// All functions are pure: no I/O, no clock reads, STDLIB only.
package validation

import (
	"fmt"
	"strings"
	"time"
)

// Severity is the outcome level of a validation rule.
type Severity string

const (
	// SeverityPass indicates the rule held.
	SeverityPass Severity = "PASS"
	// SeverityWarning indicates a non-blocking concern.
	SeverityWarning Severity = "WARNING"
	// SeverityFail indicates a blocking failure.
	SeverityFail Severity = "FAIL"
	// SeverityBlock indicates a hard block (e.g. policy not active).
	SeverityBlock Severity = "BLOCK"
)

// Rule ID constants mirroring the Python spec.
const (
	// RuleRequiredDocs requires CLAIM_FORM, HOSPITAL_FINAL_BILL, DISCHARGE_SUMMARY.
	RuleRequiredDocs = "REQUIRED_DOCS"
	// RulePolicyActive requires from <= incident <= to (when to applies).
	RulePolicyActive = "POLICY_ACTIVE"
	// RuleFieldConsistency requires normalized cross-source values to agree.
	RuleFieldConsistency = "FIELD_CONSISTENCY"
	// RuleDateLogic requires admission <= discharge.
	RuleDateLogic = "DATE_LOGIC"
	// RuleAmountReconciliation requires gross-discount == net in minor units.
	RuleAmountReconciliation = "AMOUNT_RECONCILIATION"
	// RuleDuplicateSHA256 rejects case-insensitive repeated document hashes.
	RuleDuplicateSHA256 = "DUPLICATE_SHA256"
)

// Result is the outcome of a single validation rule.
type Result struct {
	RuleID   string
	Passed   bool
	Severity Severity
	Message  string
}

func ok(ruleID, message string) Result {
	return Result{RuleID: ruleID, Passed: true, Severity: SeverityPass, Message: message}
}

func bad(ruleID, message string, severity Severity) Result {
	return Result{RuleID: ruleID, Passed: false, Severity: severity, Message: message}
}

// ValidateRequiredDocs reports which required document types are missing.
// Comparison is case-insensitive on trimmed values; blank entries are ignored.
func ValidateRequiredDocs(present, required []string) Result {
	set := make(map[string]struct{}, len(present))
	for _, p := range present {
		k := strings.ToUpper(strings.TrimSpace(p))
		if k == "" {
			continue
		}
		set[k] = struct{}{}
	}
	var missing []string
	for _, r := range required {
		k := strings.ToUpper(strings.TrimSpace(r))
		if k == "" {
			continue
		}
		if _, found := set[k]; !found {
			missing = append(missing, k)
		}
	}
	if len(missing) == 0 {
		return ok(RuleRequiredDocs, "All required documents present.")
	}
	return bad(RuleRequiredDocs, fmt.Sprintf("Missing: %s.", strings.Join(missing, ", ")), SeverityFail)
}

// ValidatePolicyActive holds iff from <= incident <= to (when hasTo).
// A zero incident or zero from, an incident before from, or an incident
// after to (when hasTo) yields a BLOCK failure.
func ValidatePolicyActive(incident, from time.Time, to time.Time, hasTo bool) Result {
	if incident.IsZero() || from.IsZero() {
		return bad(RulePolicyActive, "Incident/policy start missing.", SeverityBlock)
	}
	if incident.Before(from) {
		return bad(RulePolicyActive, "Incident precedes policy start.", SeverityBlock)
	}
	if hasTo && incident.After(to) {
		return bad(RulePolicyActive, "Incident after policy end.", SeverityBlock)
	}
	return ok(RulePolicyActive, "Policy active on incident date.")
}

// normalizeField trims, casefolds, and collapses interior whitespace.
func normalizeField(v string) string {
	return strings.Join(strings.Fields(strings.ToLower(strings.TrimSpace(v))), " ")
}

// ValidateFieldConsistency holds iff every non-empty normalized value agrees.
func ValidateFieldConsistency(field string, valuesBySource map[string]string) Result {
	distinct := make(map[string]struct{})
	for _, v := range valuesBySource {
		n := normalizeField(v)
		if n == "" {
			continue
		}
		distinct[n] = struct{}{}
	}
	if len(distinct) == 0 {
		return bad(RuleFieldConsistency, fmt.Sprintf("'%s': no values.", field), SeverityFail)
	}
	if len(distinct) == 1 {
		return ok(RuleFieldConsistency, fmt.Sprintf("'%s' consistent.", field))
	}
	return bad(RuleFieldConsistency, fmt.Sprintf("'%s' mismatch.", field), SeverityFail)
}

// ValidateDateLogic holds iff both dates are present and admission <= discharge.
func ValidateDateLogic(admission, discharge time.Time, hasAdmission, hasDischarge bool) Result {
	if !hasAdmission || !hasDischarge {
		return bad(RuleDateLogic, "Admission/discharge missing.", SeverityFail)
	}
	if admission.After(discharge) {
		return bad(RuleDateLogic, "Admission after discharge.", SeverityFail)
	}
	return ok(RuleDateLogic, "Date logic holds.")
}

// ValidateAmountPaise holds iff gross-discount == net exactly in minor units.
// Negative inputs fail; only integer arithmetic is used.
func ValidateAmountPaise(gross, discount, net int64) Result {
	if gross < 0 || discount < 0 || net < 0 {
		return bad(RuleAmountReconciliation, "Negative amount.", SeverityFail)
	}
	if gross-discount == net {
		return ok(RuleAmountReconciliation, fmt.Sprintf("Reconciled: %d-%d==%d.", gross, discount, net))
	}
	return bad(RuleAmountReconciliation, fmt.Sprintf("Expected %d, got %d.", gross-discount, net), SeverityFail)
}

// ValidateDuplicateSHA256 fails on a case-insensitive repeat of a non-empty hash.
func ValidateDuplicateSHA256(hashes []string) Result {
	seen := make(map[string]struct{}, len(hashes))
	for _, h := range hashes {
		k := strings.ToLower(strings.TrimSpace(h))
		if k == "" {
			continue
		}
		if _, dup := seen[k]; dup {
			return bad(RuleDuplicateSHA256, "Duplicate sha256.", SeverityFail)
		}
		seen[k] = struct{}{}
	}
	return ok(RuleDuplicateSHA256, "No duplicates.")
}
